### ISSUE TITLE
 Unable fo find Synthing config file', src/config.rs:63:14 - Docker

### ISSUE THREAD

**Issue by FreeSoftwareServers** [desbma/stfed#1]
Greetings,

I'm trying to get this setup as an alternative to https://github.com/terminalnetwork/syncthing-hooks as it at least offers some event triggers but I can't get it to run in docker.

I keep getting the above error and I've tried moving config files everywhere, generating them, but I'm not even sure what dir its searching as it gives a relative path.

A few notes:

* I had to add the following to top of `Config.toml`
`cargo-features = ['edition2021','strip']`
* I had to use `nightly`

Here is my `Dockerfile`:

```
cat <<'EOF'>stfed/Dockerfile
FROM liuchong/rustup:nightly

WORKDIR '/root/'
COPY git/ .
WORKDIR '/root/stfed/'

RUN cargo build --release \
    && install -Dm 755 -t /usr/local/bin target/release/stfed

EOF
chmod +x stfed/Dockerfile
```

Can you tell me what dir to place files? 

Note: SystemD isn't really available in Docker and I've been using `/usr/local/bin/stfed` to run app.

Since theres not a bunch of issues on your github, I'm going to "double dip" so we can just have one thread, but I have 2x Feature Requests:

* One is a Docker Image which accepts `Env` variables and a config file for setup, this avoids the above likely :P
* Two is that I'd like an event for "files synced remotley", this way I could run the hooks on each client. 

I'm looking at setting up a multi-client central hub setup and currently my only option is to run this on central server, but I'd like to be able to run on each client (specifically, I'm looking to delete files from client upon sync to central server).

Thanks a bunch! My next step is try installing on host to learn more, but it will need to end up containerized anyway and I started w/ a container and hoping to not install things on my host and/or setup a test VM.

**Comment by desbma** [comment]
They are at 6 least different subjects here, I'm going to answer only some of them, please open new issues for others if they matter to you.

> A few notes:
> 
>     I had to add the following to top of Config.toml
>     cargo-features = ['edition2021','strip']
>     I had to use nightly

Your Rust environment is likely outdated, this should be needed. I don't use any nightly feature.

> Can you tell me what dir to place files?

For stfed config see https://github.com/desbma/stfed#configuration
For Syncthing config see https://docs.syncthing.net/users/config.html#synopsis

**Comment by FreeSoftwareServers** [comment]
@desbma 

* That is the best/only `rustup` docker image I could find, while you may not use nightly, I wonder if it does any harm to add to your file in the repo? This would avoid anyone else having to manipulate files when pulling repo and trying to run in outdated evns? I'm new to rust, so maybe it's bad?
* I have tried moving both `config.toml` and `config.xml` to the following locations inside the container:

```
/root/.config/stfed/
/root/.config/
/root/.cargo/
/root/.cargo/bin/
/root/
/root/stfed/
/root/stfed/src/
```

It just gives the same error repeatedly?

I'll open separate threads for [FR]'s once I get this at least working, as I'll be able to give a solution to Req 1 (Docker Image) and I can work on 2 as well/live without it, but getting it running is obviously main concern.

**Comment by desbma** [comment]
> That is the best/only rustup docker image I could find, while you may not use nightly, I wonder if it does any harm to add to your file in the repo? This would avoid anyone else having to manipulate files when pulling repo and trying to run in outdated evns? I'm new to rust, so maybe it's bad?

Sorry I don't understand what you mean by "_add to your file in the repo_". If you are not familiar to Rust, I think you should use rustup and run `rustup update` to have an up to date toolchain.

>  I have tried moving both config.toml and config.xml to the following locations inside the container:

I think you should more carefully read the links I have provided. I you want to let stfed read Syncthing's config, which is the most simpler approach, the default locations are `~/.config/stfed` and `~/.config/syncthing`.

**Comment by FreeSoftwareServers** [comment]
First, thanks for the help!

What I mean is insert `cargo-features = ['edition2021','strip']` at the top of `Cargo.toml` in your GitRepo so that it's included by default when people do `git clone`.

I have used `/root/.config/stfed/config.{toml/xml)` and neither have worked. Is it perhaps that I'm not starting with `SystemD`? Any way I can pass these somehow specifically via `Env` variables>?

Have you run this in docker ever, just out of curiosity, perhaps you could share your docker build/dev>?

**Comment by desbma** [comment]
> What I mean is insert cargo-features = ['edition2021','strip'] at the top of Cargo.toml in your GitRepo so that it's included by default when people do git clone.

No, this stuff is needed in your case because you are using an outdated environment, and it has no meaning for the stable channel. Install rustup and run `rustup update` and you should be good.

> I have used /root/.config/stfed/config.{toml/xml) and neither have worked. 

Again, read the links: the Syncthing config file, if you want stfed to read it, should by default be in `~/.config/syncthing`.

**Comment by FreeSoftwareServers** [comment]
In regards to 1, yes I understand that its because I'm in an outdated Env, but perhaps I'm not the only person to have an outdated Env? Does it hurt to hard code requirements? If not, I'm just suggesting that it be added, if you don't care that others may face the same issue, fine, I've somewhat documented it here and rust output was fairly clear. The only tricky part was understanding that it gave seperate errors for `edition2021` and `strip` and I needed to combine into an array of requirements. I was able to workaround anyway and get everything to build.

In regards to 2, I did it get it to work by using `/root/.config/syncthings/config.xml`. I think my confusion arose from the documentation (which I read multiple times) that made it seem like `config.toml` could be used without the need for `sycnthings/confing.xml`, but it seems I need **both** for it to work. See --> "This file is optional,"... For me, neither file was optional.

**Comment by desbma** [comment]
> In regards to 1, yes I understand that its because I'm in an outdated Env, but perhaps I'm not the only person to have an outdated Env? Does it hurt to hard code requirements? If not, I'm just suggesting that it be added, if you don't care that others may face the same issue, fine, I've somewhat documented it here and rust output was fairly clear. The only tricky part was understanding that it gave seperate errors for edition2021 and strip and I needed to combine into an array of requirements. I was able to workaround anyway and get everything to build.

What you are missing is that the `cargo-feature` line you are using is not declaring requirement at all. It is enabling nightly features in your outdated Rust environment, for features that were not stable at the time, but that are now stable for more than a year. This line will throw an error for stable toolchains that don't allow nightly features, or older nightly that don't recognize them.
As I already said to you, the solution is simple, and is the same for all Rust projects: install rustup and run rustup update. By ignoring my advice, you are in effect still running an outdated environment, and may run into toolchain bugs that have long been fixed.

> In regards to 2, I did it get it to work by using /root/.config/syncthings/config.xml. I think my confusion arose from the documentation (which I read multiple times) that made it seem like config.toml could be used without the need for sycnthings/confing.xml, but it seems I need both for it to work. See --> "This file is optional,"... For me, neither file was optional.

* if stfed `config.toml` is present, the Syncthing config is not even read
* if you want to use the local instance, `config.toml` is optional and Syncthing config is read

I don't know how I can state any clearer than quoting the README:

"_This file is optional, if absent, stfed will read Syncthing configuration and build its own configuration from it. You typically only need to use this file if using a non local Syncthing instance._"

**Comment by FreeSoftwareServers** [comment]
In regards to 1, I am new to rust and didn't understand, but that makes sense then if it will cause errors. I will search for a more updated `rustup` base image or perhaps have to build one from a base OS. This is a great reason that a provided docker image would be useful.

In regards to "I don't know how I can state any clearer than quoting the README:"

I am trying to clearly state that if I removed either `/root/.config/stfed/config.toml` OR `/root/.config/syncthing/config.xml` then it failed to launch. Neither file was optional, but BOTH were REQUIRED.

**Comment by desbma** [comment]
> I am trying to clearly state that if I removed either /root/.config/stfed/config.toml OR /root/.config/syncthing/config.xml then it failed to launch. Neither file was optional, but BOTH were REQUIRED.

Which I highly doubt, since I am running `stfed` without `~/.config/stfed/config.toml` myself.

Maybe you are confusing with the hooks file, which is mandatory ?

**Comment by FreeSoftwareServers** [comment]
> > I am trying to clearly state that if I removed either /root/.config/stfed/config.toml OR /root/.config/syncthing/config.xml then it failed to launch. Neither file was optional, but BOTH were REQUIRED.
> 
> Which I highly doubt, since I am running `stfed` without `~/.config/stfed/config.toml` myself.
> 
> Maybe you are confusing with the hooks file, which is mandatory ?

I can confirm as below and can share how to reproduce if you're interested.

Note that my config.xml has 127 as it's running in a container, but this works fine.

In essence, without `config.toml` it will use `config.xml`, this only failed because the IP was wrong and likely would work if I adjusted IP in config.

But, if I remove `config.xml` then I get the error this all started with, so in essence, I can't use `config.toml` without `config.xml`.

It is however reading my `config.toml` as when I have both files, I can leave the 127 address in the `config.xml` and it uses the IP provided in `config.toml`.

Remove `config.xml` =

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Unable to guess stfed configuration field values from Synthing config, please write a config file

Caused by:
    Unable fo find Synthing config file', src/config.rs:63:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Remove `config.toml` = 

```
WARN  [stfed::config] Unable to find config file, using default config
WARN  [stfed] Syncthing·server·connection failed,·will·restart·main·loop.·http://127.0.0.1:8384/rest/system/config: Connection Failed: Connect error: Connection refused (os error 111)

Caused by:
    Connection refused (os error 111)
INFO  [stfed] Will reconnect in 5s
```

With Both Files = 

`INFO  [stfed] New event: FolderDownSyncDone { folder: "/var/syncthing/Sync" }`

If you're intertested in reproducding:

Here is how I setup syncthings, all default currently --> https://www.freesoftwareservers.com/display/FREES/SyncThings+-+Docker-Compose

Here is how I added/tested `stfed` --> https://www.freesoftwareservers.com/display/FREES/SyncThings+-+desbma+-+Hooks+-+Docker-Compose

**Comment by desbma** [comment]
> Remove `config.toml` =
> 
> ```
> WARN  [stfed::config] Unable to find config file, using default config
> WARN  [stfed] Syncthing·server·connection failed,·will·restart·main·loop.·http://127.0.0.1:8384/rest/system/config: Connection Failed: Connect error: Connection refused (os error 111)
> 
> Caused by:
>     Connection refused (os error 111)
> INFO  [stfed] Will reconnect in 5s
> ```

This error is not due to any missing config file. Either Syncthing is not running, or it is not listening on the configured IP/port.

**Comment by FreeSoftwareServers** [comment]
I'm not sure if you're a reading what I'm saying as I already clearly stated that.

The issue is I couldn't use config.toml without config.xml present. Did you ever actually test this standalone setup?

I purposefully didn't want/try to use config.xml as I knew it had 127 address.

**Comment by desbma** [comment]
It took me a while to understand your issue, but I think I get it now.

**Comment by desbma** [comment]
Please test 93633b2c778b0953f1dd07302d86f78180bdc5be

**Comment by FreeSoftwareServers** [comment]
> Please test [93633b2](https://github.com/desbma/stfed/commit/93633b2c778b0953f1dd07302d86f78180bdc5be)

I'll test later at home, appreciate the feedback. I didn't fully understand the problem until I had resolved it tbh. My assumption/thought is that I could have even done `touch /root/.config/syncthing/config.xml` and it would have used the `config.toml` but it was failing due to not finding `config.xml` and wasn't getting to the part where it looked for `config.toml` in the code. As I mentioned before, new to rust, I did read your code, but it wasn't easy for me to read like a NodeJS project would be.

Thanks for your time, I'll be sure to test and let you know.

**Comment by FreeSoftwareServers** [comment]
> Please test [93633b2](https://github.com/desbma/stfed/commit/93633b2c778b0953f1dd07302d86f78180bdc5be)

Confirmed working. I also wanted to mention I was able to use "stable" and not add lines by using the official Rust docker image. I thought rustup was additional software that needed to be installed,  so I used an image tagged "rustup", but its included in base Rust image, so the official image worked. 

I was thinking to put in a FR to be able to configure a command for each event in the same hooks.toml. IE: I'll likely want either file/folder sync complete and file sync error (so 2 seperate commands) and it would be nice to not have to run 2x containers, but I def can just do that.

Thanks for reso!
